cmd/zpool/zpool_main.c

```diff
@@ -2844,12 +3441,262 @@ get_timestamp_arg(char c)
 }
 
 /*
- * zpool iostat [-gLPv] [-T d|u] [pool] ... [interval [count]]
+ * Return stat flags that are supported by all pools.  "data" should be
+ * initialized to all 0xFFs before running.  It will get ANDed down until
+ * only the flags that are supported on all pools remain.
+ */
+static int
+get_stat_flags_cb(zpool_handle_t *zhp, void *data)
+{
+	uint64_t *mask = data;
+	nvlist_t *config, *nvroot;
+	uint64_t flags = 0;
+	int i, j;
+
+	/*
+	 * Lookup table for iostat flags to nvlist names.  Basically a list
+	 * of all the nvlists a flag requires.
+	 */
+	static const char *type_to_nvlist[IOS_COUNT][10] = {
+		[IOS_DEFAULT] = {ZPOOL_CONFIG_VDEV_STATS, NULL},
+		[IOS_L_HISTO] = {ZPOOL_CONFIG_TOT_R_LAT_HISTO,
+		    ZPOOL_CONFIG_TOT_W_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_R_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_W_LAT_HISTO,
+		    ZPOOL_CONFIG_SYNC_R_LAT_HISTO,
+		    ZPOOL_CONFIG_SYNC_W_LAT_HISTO,
+		    ZPOOL_CONFIG_ASYNC_R_LAT_HISTO,
+		    ZPOOL_CONFIG_ASYNC_W_LAT_HISTO,
+		    ZPOOL_CONFIG_SCRUB_LAT_HISTO, NULL},
+		[IOS_LATENCY] = {ZPOOL_CONFIG_TOT_R_LAT_HISTO,
+		    ZPOOL_CONFIG_TOT_W_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_R_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_W_LAT_HISTO, NULL},
+		[IOS_QUEUES] = {ZPOOL_CONFIG_QUEUE_ACTIVE, NULL}
+	};
+
+	config = zpool_get_config(zhp, NULL);
+	verify(nvlist_lookup_nvlist(config, ZPOOL_CONFIG_VDEV_TREE,
+	    &nvroot) == 0);
+
+	/* For each type, make sure all its nvlist entries are supported */
+	for (j = 0; j < ARRAY_SIZE(type_to_nvlist); j++) {
+		if (!type_to_nvlist[j][0])
+			continue;
+
+		/* Start off by assuming the flag is supported, then check */
+		flags |= (1ULL << j);
+		for (i = 0; type_to_nvlist[j][i]; i++) {
+			if (!nvlist_exists(nvroot, type_to_nvlist[j][i])) {
+				/* flag isn't supported */
+				flags = flags & ~(1ULL  << j);
+				break;
+			}
+		}
+	}
+
+	*mask = *mask & flags;
+	return (0);
+}
+
+/*
+ * Return a bitmask of stats that are supported on all pools.
+ */
+static uint64_t
+get_stat_flags(zpool_list_t *list)
+{
+	uint64_t mask = -1;
```

Comment:
`get_stat_flags()` will return a bitmask of the IOS_* bits which are supported by both the client and module (see `get_stat_flags_cb()`).  I need to update the function comments to make that clear though; right now it still has the old comment from when `get_stats_flags()` was strictly returning the bits supported by the module only.

In interim versions of the code I did have `mask = VS_CURRENT_FLAGS`, but later removed it when I switched to using per-nvlist histograms (which removed the VS_CURRENT_FLAGS macro since it was no longer needed).  VS_CURRENT_FLAGS was basically the IOS_ALLFLAGS_M mask you described.